src/api/Feed.js

```diff
@@ -0,0 +1,1022 @@
+     * @param {Array} items - the feed items to cache
+     */
+    setCachedItems(id: string, items: FeedItems) {
+        const cache = this.getCache();
+        const cacheKey = this.getCacheKey(id);
+        cache.set(cacheKey, {
+            hasError: !!this.hasError,
+            items
+        });
+    }
+
+    /**
+     * Gets the feed items
+     *
+     * @param {BoxItem} file - The file to which the task is assigned
+     * @param {Function} successCallback - the success callback
+     * @param {Function} errorCallback - the error callback
+     */
+    feedItems(file: BoxItem, successCallback: Function, errorCallback: Function): void {
+        const { id } = file;
+
+        if (!id) {
+            throw getBadItemError();
+        }
+
+        const cachedItems = this.getCachedItems(id);
+        if (cachedItems) {
+            const { hasError, items } = cachedItems;
+            if (hasError) {
+                errorCallback(items);
+            } else {
+                successCallback(items);
+            }
+
+            return;
+        }
+
+        this.id = id;
+        this.hasError = false;
+        const versionsPromise = this.fetchVersions();
+        const commentsPromise = this.fetchComments();
+        const tasksPromise = this.fetchTasks();
+
+        Promise.all([versionsPromise, commentsPromise, tasksPromise]).then((feedItems) => {
+            const versions: ?FileVersions = feedItems[0];
+            const versionsWithRestoredVersion = this.addCurrentVersion(versions, file);
+            const unsortedFeedItems = [versionsWithRestoredVersion, ...feedItems.slice(1)];
+            const sortedFeedItems = this.sortFeedItems.apply(null, unsortedFeedItems);
+            this.setCachedItems(id, sortedFeedItems);
+            if (!this.isDestroyed()) {
+                if (this.hasError) {
+                    errorCallback(sortedFeedItems);
+                } else {
+                    successCallback(sortedFeedItems);
+                }
+            }
+        });
+    }
+
+    /**
+     * Sort valid feed items, descending by created_at time.
+     *
+     * @param {Array<?Comments | ?Tasks | ?FileVersions>} args - Arguments list of each item container
+     * type that is allowed in the feed.
+     * @return {Array<?Comments | ?Tasks | ?FileVersions>} the sorted feed items
+     */
+    sortFeedItems(...args: Array<?Comments | ?Tasks | ?FileVersions>): FeedItems {
+        const feedItems = [];
+        args.forEach((itemContainer) => {
+            if (itemContainer) {
+                feedItems.push(...itemContainer.entries);
+            }
+        });
+        return feedItems.sort((a, b) => Date.parse(b.created_at) - Date.parse(a.created_at));
+    }
+
+    /**
+     * Fetches the comments for a file
+     *
+     * @return {Promise} - the file comments
+     */
+    fetchComments(): Promise<?Comments> {
+        this.commentsAPI = new CommentsAPI(this.options);
+        return new Promise((resolve) => {
```

Comment:
no async/await?

Reply:
Should the various component services return a promise directly rather than wrapping it at the consumer level? Or is that not possible due to async/await being used in the low-level modules?

Reply:
Lower level API's use callbacks, so thats why I wrapped it in a promise for convenience